syntax = "proto3";

import "google/protobuf/empty.proto";
import "google/protobuf/struct.proto";

package specter_proto;

// -------------------------------- Previewer ------------------------------- //

service PreviewerService {
    rpc ListenPreview (Object) returns (stream PreviewImage) {}
}

message PreviewImage {
    bytes image = 1;
}

// --------------------------------- Marker --------------------------------- //

service MarkerService {
    rpc Start (google.protobuf.Empty) returns (google.protobuf.Empty) {}
    rpc Stop (google.protobuf.Empty) returns (google.protobuf.Empty) {}
}

// --------------------------------- Recorder -------------------------------- //

service RecorderService {
    rpc ListenCommands (google.protobuf.Empty) returns (stream RecorderCommand) {}
}

message RecorderCommand {
    string command = 1;
}

// --------------------------------- Object ---------------------------------- //

service ObjectService {
    rpc GetTree (OptionalObject) returns (ObjectTree) {}
    rpc Find (Object) returns (Objects) {}

    rpc GetParent (Object) returns (Object) {}
    rpc GetChildren (Object) returns (Objects) {}

    rpc CallMethod (MethodCall) returns (google.protobuf.Empty) {}
    rpc UpdateProperty (PropertyUpdate) returns (google.protobuf.Empty) {}

    rpc GetMethods (Object) returns (Methods) {}
    rpc GetProperties (Object) returns (Properties) {}

    rpc ListenTreeChanges (google.protobuf.Empty) returns (stream ObjectChange) {}
    rpc ListenPropertiesChanges (Object) returns (stream PropertyChange) {}
}

message OptionalObject {
    optional string query = 1;
}

message Object {
    string query = 1;
}

message Objects {
    repeated Object objects = 1;
}

message ObjectTree {
    repeated ObjectNode roots = 1;
}

message ObjectNode {
    Object object = 1;
    repeated ObjectNode children = 2;
}

message MethodCall {
    Object object = 1;
    string method = 2;
    repeated google.protobuf.Value arguments = 3;
}

message PropertyUpdate {
    Object object = 1;
    string property = 2;
    google.protobuf.Value value = 3;
}

message Methods {
    repeated Method methods = 1;
}

message Method {
    string name = 1;
    repeated Parameter parameters = 2;
}

message Parameter {
    string name = 1;
    google.protobuf.Value default_value = 2;
}

message Properties {
    repeated Property properties = 1;
}

message Property {
    string name = 1;
    google.protobuf.Value value = 2;
    bool read_only = 3;
}

message ObjectChange {
    oneof change_type {
        ObjectAdded added = 1;
        ObjectRemoved removed = 2;
        ObjectReparented reparented = 3;
        ObjectRenamed renamed = 4;
    }
}

message ObjectAdded {
    Object object = 1;
    Object parent = 2;
}

message ObjectRemoved {
    Object object = 1;
}

message ObjectReparented {
    Object object = 1;
    Object parent = 2;
}

message ObjectRenamed {
    Object old_object = 1;
    Object new_object = 2;
}

message PropertyChange {
    oneof change_type {
        PropertyAdded added = 1;
        PropertyRemoved removed = 2;
        PropertyUpdated updated = 4;
    }
}

message PropertyAdded {
    string property = 1;
    google.protobuf.Value value = 2;
    bool read_only = 3;
}

message PropertyRemoved {
    string property = 1;
}

message PropertyUpdated {
    string property = 1;
    google.protobuf.Value old_value = 2;
    google.protobuf.Value new_value = 3;
}
